### PR TITLE
Add new commands and event handlers introduced in zwave-js 8.9.0

### PR DESCRIPTION
**Pull request by raman325**
In this PR we add the following node level commands as well as the progress events that they emit:
- `test_power_level`
- `check_lifeline_health`
- `check_route_health`

Since no existing commands changed, I kept the minimum schema version to 12 even though I bumped up the max to 13. With that being said, if we want to force users to update to the latest to address that options initialization issue, we can bump the min version as well

### CODE REVIEW

**Comment by MartinHjelmare** [comment]
If we want to start using the new commands in Home Assistant we need to bump minimum schema version.

File: zwave_js_server/const/__init__.py [hunk 1]
```diff
@@ -248,3 +248,19 @@ class ZwaveFeature(IntEnum):
 
     # https://github.com/zwave-js/node-zwave-js/blob/master/packages/zwave-js/src/lib/controller/Features.ts#L4
     SMART_START = 0
+
+
+class Powerlevel(IntEnum):
```
Comment (MartinHjelmare) [inline zwave_js_server/const/__init__.py:253]:
```suggestion
class PowerLevel(IntEnum):
```

File: zwave_js_server/model/node.py [hunk 2]
```diff
@@ -97,6 +104,192 @@ def timeout_response(self) -> int:
...
+    def route_changes(self) -> Optional[int]:
+        """Return number of route changes."""
+        return self.data.get("routeChanges")
+
+    @property
+    def min_powerlevel(self) -> Optional[Powerlevel]:
+        """Return minimum powerlevel."""
+        powerlevel = self.data.get("minPowerlevel")
+        if powerlevel is not None:
+            return Powerlevel(powerlevel)
```
Comment (MartinHjelmare) [inline zwave_js_server/model/node.py:160]:
```suggestion
        power_level = self.data.get("minPowerlevel")
        if power_level is not None:
            return PowerLevel(power_level)
```

File: zwave_js_server/model/node.py [hunk 3]
```diff
@@ -1,7 +1,14 @@
 """Provide a model for the Z-Wave JS node."""
+from dataclasses import dataclass
 from typing import TYPE_CHECKING, Any, Dict, List, Optional, TypedDict, Union, cast
 
-from ..const import INTERVIEW_FAILED, CommandClass, NodeStatus, SecurityClass
+from ..const import (
+    INTERVIEW_FAILED,
+    CommandClass,
+    NodeStatus,
+    Powerlevel,
```
Comment (MartinHjelmare) [inline zwave_js_server/model/node.py:9]:
```suggestion
    PowerLevel,
```

File: zwave_js_server/model/node.py [hunk 4]
```diff
@@ -97,6 +104,192 @@ def timeout_response(self) -> int:
...
+        if powerlevel is not None:
+            return Powerlevel(powerlevel)
+        return None
+
+    @property
+    def min_powerlevel_target(self) -> Optional[Powerlevel]:
+        """Return minimum powerlevel target."""
+        powerlevel = self.data.get("minPowerlevelTarget")
+        if powerlevel is not None:
+            return Powerlevel(powerlevel)
```
Comment (MartinHjelmare) [inline zwave_js_server/model/node.py:253]:
```suggestion
    def min_power_level_target(self) -> Optional[PowerLevel]:
        """Return minimum power level target."""
        power_level = self.data.get("minPowerlevelTarget")
        if power_level is not None:
            return PowerLevel(power_level)
```

File: zwave_js_server/model/node.py [hunk 5]
```diff
@@ -97,6 +104,192 @@ def timeout_response(self) -> int:
...
+    def failed_pings_to_source(self) -> Optional[int]:
+        """Return number of failed pings to source."""
+        return self.data.get("failedPingsToSource")
+
+    @property
+    def min_powerlevel_source(self) -> Optional[Powerlevel]:
+        """Return minimum powerlevel source."""
+        powerlevel = self.data.get("minPowerlevelSource")
+        if powerlevel is not None:
+            return Powerlevel(powerlevel)
```
Comment (MartinHjelmare) [inline zwave_js_server/model/node.py:245]:
```suggestion
    def min_power_level_source(self) -> Optional[PowerLevel]:
        """Return minimum power level source."""
        power_level = self.data.get("minPowerlevelSource")
        if power_level is not None:
            return PowerLevel(power_level)
```

File: zwave_js_server/model/node.py [hunk 6]
```diff
@@ -539,6 +732,71 @@ async def async_get_highest_security_class(self) -> SecurityClass:
         assert data
         return SecurityClass(data["highestSecurityClass"])
 
+    async def async_test_powerlevel(
+        self, test_node_id: int, powerlevel: Powerlevel, test_frame_count: int
```
Comment (MartinHjelmare) [inline zwave_js_server/model/node.py:736]:
```suggestion
    async def async_test_power_level(
        self, test_node_id: int, power_level: PowerLevel, test_frame_count: int
```

File: zwave_js_server/model/node.py [hunk 7]
```diff
@@ -539,6 +732,71 @@ async def async_get_highest_security_class(self) -> SecurityClass:
...
+    ) -> int:
+        """Send testPowerLevel command to Node."""
+        data = await self.async_send_command(
+            "test_powerlevel",
+            testNodeId=test_node_id,
+            powerlevel=powerlevel,
```
Comment (MartinHjelmare) [inline zwave_js_server/model/node.py:742]:
```suggestion
            powerlevel=power_level,
```

File: zwave_js_server/model/node.py [hunk 8]
```diff
@@ -539,6 +732,71 @@ async def async_get_highest_security_class(self) -> SecurityClass:
...
+            require_schema=13,
+            wait_for_result=True,
+            **kwargs,
+        )
+        assert data
+        return RouteHealthCheckSummary(data["summary"])
+
+    def handle_test_powerlevel_progress(self, event: Event) -> None:
+        """Process a test powerelevel progress event."""
+        event.data["test_powerlevel_progress"] = TestPowerlevelProgress(
```
Comment (MartinHjelmare) [inline zwave_js_server/model/node.py:784]:
```suggestion
    def handle_test_power_level_progress(self, event: Event) -> None:
        """Process a test power level progress event."""
        event.data["test_power_level_progress"] = TestPowerLevelProgress(
```

File: test/model/test_node.py [hunk 9]
```diff
@@ -8,6 +8,7 @@
     INTERVIEW_FAILED,
     CommandClass,
     NodeStatus,
+    Powerlevel,
```
Comment (MartinHjelmare) [inline test/model/test_node.py:11]:
```suggestion
    PowerLevel,
```

File: test/model/test_node.py [hunk 10]
```diff
@@ -998,3 +1004,168 @@ async def test_get_highest_security_class(multisensor_6: Node, uuid4, mock_comma
         "nodeId": node.node_id,
         "messageId": uuid4,
     }
+
+
+async def test_test_powerlevel(multisensor_6: Node, uuid4, mock_command):
```
Comment (MartinHjelmare) [inline test/model/test_node.py:1009]:
```suggestion
async def test_test_power_level(multisensor_6: Node, uuid4, mock_command):
```

File: test/model/test_node.py [hunk 11]
```diff
@@ -998,3 +1004,168 @@ async def test_get_highest_security_class(multisensor_6: Node, uuid4, mock_comma
...
+    """Test node.test_powerlevel command."""
+    node = multisensor_6
+    ack_commands = mock_command(
+        {"command": "node.test_powerlevel", "nodeId": node.node_id},
+        {"framesAcked": 1},
+    )
+    assert await node.async_test_powerlevel(2, Powerlevel.DBM_MINUS_1, 3) == 1
```
Comment (MartinHjelmare) [inline test/model/test_node.py:1016]:
```suggestion
    assert await node.async_test_power_level(2, Powerlevel.DBM_MINUS_1, 3) == 1
```

File: test/model/test_node.py [hunk 12]
```diff
@@ -998,3 +1004,168 @@ async def test_get_highest_security_class(multisensor_6: Node, uuid4, mock_comma
         "nodeId": node.node_id,
         "messageId": uuid4,
     }
+
+
+async def test_test_powerlevel(multisensor_6: Node, uuid4, mock_command):
+    """Test node.test_powerlevel command."""
```
Comment (MartinHjelmare) [inline test/model/test_node.py:1010]:
```suggestion
    """Test node.test_power_level command."""
```

File: test/model/test_node.py [hunk 13]
```diff
@@ -998,3 +1004,168 @@ async def test_get_highest_security_class(multisensor_6: Node, uuid4, mock_comma
...
+        {"framesAcked": 1},
+    )
+    assert await node.async_test_powerlevel(2, Powerlevel.DBM_MINUS_1, 3) == 1
+
+    assert len(ack_commands) == 1
+    assert ack_commands[0] == {
+        "command": "node.test_powerlevel",
+        "nodeId": node.node_id,
+        "testNodeId": 2,
+        "powerlevel": Powerlevel.DBM_MINUS_1.value,
```
Comment (MartinHjelmare) [inline test/model/test_node.py:1023]:
```suggestion
        "powerlevel": PowerLevel.DBM_MINUS_1.value,
```

File: test/model/test_node.py [hunk 14]
```diff
@@ -998,3 +1004,168 @@ async def test_get_highest_security_class(multisensor_6: Node, uuid4, mock_comma
...
+        "testFrameCount": 3,
+        "messageId": uuid4,
+    }
+
+
+async def test_test_powerlevel_progress_event(multisensor_6: Node, uuid4, mock_command):
```
Comment (MartinHjelmare) [inline test/model/test_node.py:1029]:
```suggestion
async def test_test_power_level_progress_event(multisensor_6: Node, uuid4, mock_command):
```

File: test/model/test_node.py [hunk 15]
```diff
@@ -998,3 +1004,168 @@ async def test_get_highest_security_class(multisensor_6: Node, uuid4, mock_comma
...
+    """Test test powerlevel progress event."""
```
Comment (MartinHjelmare) [inline test/model/test_node.py:1030]:
```suggestion
    """Test test power level progress event."""
```

File: test/model/test_node.py [hunk 16]
```diff
@@ -998,3 +1004,168 @@ async def test_get_highest_security_class(multisensor_6: Node, uuid4, mock_comma
...
+            "event": "test powerlevel progress",
+            "nodeId": multisensor_6.node_id,
+            "acknowledged": 1,
+            "total": 2,
+        },
+    )
+    multisensor_6.receive_event(event)
+    assert event.data["test_powerlevel_progress"]
+    assert event.data["test_powerlevel_progress"].acknowledged == 1
+    assert event.data["test_powerlevel_progress"].total == 2
```
Comment (MartinHjelmare) [inline test/model/test_node.py:1044]:
```suggestion
    assert event.data["test_power_level_progress"]
    assert event.data["test_power_level_progress"].acknowledged == 1
    assert event.data["test_power_level_progress"].total == 2
```

File: test/model/test_node.py [hunk 17]
```diff
@@ -998,3 +1004,168 @@ async def test_get_highest_security_class(multisensor_6: Node, uuid4, mock_comma
...
+        },
+    )
+    summary = await node.async_check_lifeline_health(1)
+
+    assert summary.rating == 10
+    assert summary.results[0].latency == 1
+    assert summary.results[0].num_neighbors == 2
+    assert summary.results[0].failed_pings_node == 3
+    assert summary.results[0].route_changes == 4
+    assert summary.results[0].min_powerlevel == Powerlevel.DBM_MINUS_5
```
Comment (MartinHjelmare) [inline test/model/test_node.py:1076]:
```suggestion
    assert summary.results[0].min_power_level == PowerLevel.DBM_MINUS_5
```

File: test/model/test_node.py [hunk 18]
```diff
@@ -998,3 +1004,168 @@ async def test_get_highest_security_class(multisensor_6: Node, uuid4, mock_comma
...
+    )
+    summary = await node.async_check_route_health(15, 1)
+
+    assert summary.rating == 10
+    assert summary.results[0].num_neighbors == 1
+    assert summary.results[0].rating == 10
+    assert summary.results[0].failed_pings_to_source == 2
+    assert summary.results[0].failed_pings_to_target == 3
+    assert summary.results[0].min_powerlevel_source == Powerlevel.DBM_MINUS_4
+    assert summary.results[0].min_powerlevel_target == Powerlevel.DBM_MINUS_5
```
Comment (MartinHjelmare) [inline test/model/test_node.py:1140]:
```suggestion
    assert summary.results[0].min_power_level_source == PowerLevel.DBM_MINUS_4
    assert summary.results[0].min_power_level_target == PowerLevel.DBM_MINUS_5
```

File: zwave_js_server/model/node.py [hunk 19]
```diff
@@ -97,6 +104,192 @@ def timeout_response(self) -> int:
...
+        return self.data["failedPingsNode"]
+
+    @property
+    def route_changes(self) -> Optional[int]:
+        """Return number of route changes."""
+        return self.data.get("routeChanges")
+
+    @property
+    def min_powerlevel(self) -> Optional[Powerlevel]:
+        """Return minimum powerlevel."""
```
Comment (MartinHjelmare) [inline zwave_js_server/model/node.py:157]:
```suggestion
        """Return minimum power level."""
```

**Review by MartinHjelmare — APPROVED** [review]

**Comment by raman325** [comment]
so I had originally written out power level as two words, however `zwave-js` treats it as one: https://zwave-js.github.io/node-zwave-js/#/api/node?id=testpowerlevel

Should we do what seems most logical (break it into two words) or stay as close to the API conventions as possible (one word)?

**Comment by MartinHjelmare** [comment]
I think we should use two words in our API. That's what all the change suggestions have done. I would consider it a typo in zwave-js, so if they fix that we don't need to change our API.

File: test/model/test_node.py (hunk 12 above)
Comment (raman325) [reply test/model/test_node.py:1010]: this is incorrect, the command on the server side is still `node.test_powerlevel`

File: zwave_js_server/model/node.py (hunk 8 above)
Comment (raman325) [reply zwave_js_server/model/node.py]: this function name change is incorrect, the event is currently `test powerlevel progress`

**Comment by raman325** [comment]
These commands were removed in a later zwave-js release. We may end up throwing an error in these commands on the server and reintroduce them in a future schema version once the commands have been added back in

**Comment by raman325** [comment]
With 8.9.0 now supported in the server we can merge this when approved

**Review by MartinHjelmare — APPROVED** [review]

**Pull request merged**